Generics checks for parameter passing

NikitaAware opened the pull request:
Thank you for contributing to NullAway!

Please note that once you click "Create Pull Request" you will be asked to sign our [Uber Contributor License Agreement](https://cla-assistant.io/uber/NullAway) via [CLA assistant](https://cla-assistant.io/).

Before pressing the "Create Pull Request" button, please provide the following:

  - [x] A description about what and why you are contributing, even if it's trivial.

  - [x] The issue number(s) or PR number(s) in the description if you are contributing in response to those.

  - [x] If applicable, unit tests.

NikitaAware committed:
test

NikitaAware committed:
test

NikitaAware committed:
method

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -422,4 +442,27 @@ public void checkTypeParameterNullnessForAssignabilityForConditionalExpression(
+    }
+    // Check if the parameters are generics and compare annotations for the actual and formal
+    // parameters
+    for (int i = 0; i < formalParams.size(); i++) {
```
We should add a test case where the invoked method has varargs, like these:

https://github.com/uber/NullAway/blob/master/nullaway/src/test/java/com/uber/nullaway/NullAwayVarargsTests.java#L47

We should write something like:

```java
void foo(NullableTypeParam<@Nullable String>... args) { ... }
```
And then _all_ the parameters passed at a call to `foo()` should have `@Nullable String` as their type parameter.

NikitaAware replied:
Thanks. I will update this.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:
```diff
@@ -177,6 +177,26 @@ private static void reportMismatchedTypeForTernaryOperator(
+    ErrorBuilder errorBuilder = analysis.getErrorBuilder();
+    ErrorMessage errorMessage =
+        new ErrorMessage(
+            ErrorMessage.MessageTypes.TYPE_PARAMETER_CANNOT_BE_NULLABLE,
```
we'll want a new type of error here

NikitaAware replied:
something like GNERIC_TYPE_PARAMETER_CANNOT_BE_NULLABLE ?

msridhar replied:
Maybe `PASS_NULLABLE_GENERIC`

msridhar reviewed (commented):
couple of quick comments

NikitaAware committed:
updated test

NikitaAware committed:
updated test

msridhar commented on nullaway/src/test/java/com/uber/nullaway/NullAwayJSpecifyGenericsTests.java:
```diff
@@ -632,6 +628,35 @@ public void nestedTernary() {
+            "  static A<String> sampleMethod(A<A<String>> a1, A<String> a2) {",
+            "     return a2;",
+            "  }",
+            "  static A<@Nullable String> sampleMethodWithVarArgs(A<A<@Nullable String>> a1, A<@Nullable String> a2, @Nullable Object o1, @Nullable Object o2) {",
```
This is not using varargs.  You can read about varargs here:

https://www.baeldung.com/java-varargs

NikitaAware committed:
new error message type

NikitaAware committed:
vararg support

msridhar committed:
tweak error message

msridhar committed:
separate out varargs test

msridhar committed:
improve test

msridhar committed:
correct varargs check

msridhar committed:
Merge branch 'generics-ternary-operator-assignment-changes' into generics-checks-for-parameter-passing

NikitaAware committed:
minor changes in a test case

nimakarimipour committed:
Serialization version 3: update method serialization to exclude type use annotations and type arguments (#735)

This PR updates serialization to version `3` and burns version `2`.

Changes in serialization version 3:

- Type arguments and Type use annotations are excluded from the serialized method signatures.

---
Additional context:
This PR updates method serialization to ensure only relevant information in a method signature is serialized. Previous to this PR, method signatures serialization was relying on calling  `.toString()` method, but if a parameter contains an annotation of type `@Target({TYPE_USE})` the annotation will exist in the `.toSting()` output which is not part of a method signature. This PR updates serialization to ensure exact method signature as expected is serialized.

msridhar committed:
Merge branch 'generics-ternary-operator-assignment-changes' into generics-checks-for-parameter-passing

msridhar reviewed (approved):
LGTM!  Let's hold off for the moment on opening a PR upstream

ybroeker committed:
Docs: `-XepExcludedPaths` was added in 2.1.3, not 2.13 (#744)

lazaroclapp committed:
Add command line option to skip specific library models. (#741)

This PR adds the CLI option `-XepOpt:NullAway:IgnoreLibraryModelsFor` which takes a list of methods, given as fully qualified class name + method simple name (i.e. independent of argument types).

Methods matching this list will be skipped when loading library models (from any implementation of the `LibraryModels`'s `@AutoService` as well as our included models). This can be used in a per-compilation target basis to disable NullAway's library models for ease of upgrading to new versions with stricter modeling of common JDK or third-party APIs.

We considered a few alternative approaches here, but decided against:

- Simply using another instance of `LibraryModels` to "invert" the models given by NullAway's default library models: This would have required no code changes, but doesn't work in all cases. If NullAway's default models make the return of method `foo()` `@Nullable`, for example, then a new model that makes the return `@NonNull` will break `@Nullable` overrides of `foo()`. In general, we want to go back to "optimistic assumptions" rather than just replace the library model.
- We could have a list of methods in the `LibraryModels` for which to ignore previous models, and have those override any models on those methods coming from a different `LibraryModels` implementation. But, from the point of view of the user configuring NullAway, this is complex: they need to have an instance of custom library models in their build, and changing java plugin classpath deps on a per-target basis is more complex than changing CLI arguments (e.g. due to JVM re-use by the build system).
- We could provide more specific disabling of library models (e.g. a specific method signature or removing only one particular kind of model from a method, such as keeping the model on the return value, but removing it from an argument, or removing a null-implies-false model or similar). We could revisit this in the future, but supporting this would make the syntax of the CLI values a lot more complex. For now, we believe just turning off all models for a given method is a sufficient degree of granularity.
- Per-package/per-class/regex based ignore specs: See above. Avoiding complexity until we need it.

Note: If and when this lands, it needs a Wiki documentation update!

---------
Co-authored-by: Manu Sridharan <msridhar@gmail.com>

lazaroclapp committed:
Prepare for release 0.10.10.

lazaroclapp committed:
Prepare next development version.

msridhar committed:
Update to Gradle 8.0.2 (#743)

We had to update a couple other plugins and fix a couple of things in our Gradle build scripts to be compatible with Gradle 8.

XN137 committed:
[Refactor] Pass resolved Symbols into Handler methods (#729)

This avoids redundant ASTHelpers.getSymbol calls and is more consistent with the other Handler method signatures

NikitaAware committed:
JSpecify generics checks for conditional expressions (#739)

For a conditional expression _e_, the type parameter nullability annotations of the true-subpart and false-subpart should match the annotations required by _e_'s context.  E.g., if _e_ is the right-hand side of an assignment, the annotations should match those for the type parameters for the assignment's left-hand side:
```java
// this is fine
A<@Nullable String> t = condition? new A<@Nullable String> : new A<@Nullable String>();
// this is bad
A<@Nullable String> t2 = condition? new A<String> : new A<String>();
// this is also bad
A<@Nullable String>  t1 = condition ? new A<String>(): new A<@Nullable String>()
```
Other contexts include returns, parameter passing, and being nested inside another conditional expression.  In our experience, it seems that `javac` captures the type parameter nullability required by _e_'s context in the type of _e_ itself.  So, our handling of conditional expressions simply checks that the types of both the true and false sub-expressions of _e_ are assignable to (i.e., a subtype of) the type of _e_, using the same machinery for checking assignments from #715.

XN137 committed:
Prepare for Nullable `ASTHelpers.getSymbol` (#733)

This method will return `@Nullable` in the next Error Prone release: https://github.com/google/error-prone/issues/3760 This change prepares for that by adding appropriate null checks.

msridhar committed:
Merge branch 'master' into generics-checks-for-parameter-passing
